File: src/1.UnionFind/WeightedQuickUnionUF.cpp
#include <vector>

class WeightedQuickUnionUF {
private:
    std::vector<int> id;
    std::vector<int> sz;

    int root(int i) {
        while (i != id[i]) {
            // Path compression: make node point to its grandparent
            id[i] = id[id[i]];
            i = id[i];
        }
        return i;
    }

public:
    WeightedQuickUnionUF(int N) {
        id.resize(N);
        sz.resize(N, 1);
        for (int i = 0; i < N; ++i) {
            id[i] = i;
        }
    }

    bool connected(int p, int q) {
        return root(p) == root(q);
    }

    void Union(int p, int q) {  // Renamed from 'union' (which is a keyword)
        int i = root(p);
        int j = root(q);
        if (i == j) return;

        // Attach smaller tree to larger tree
        if (sz[i] < sz[j]) {
            id[i] = j;
            sz[j] += sz[i];
        }
        else {
            id[j] = i;
            sz[i] += sz[j];
        }
    }
};

int main() {
    WeightedQuickUnionUF uf(10);
    uf.Union(4, 3);
    uf.Union(3, 8);
    uf.Union(6, 5);
    uf.Union(9, 4);

    bool connected = uf.connected(8, 9); // Should be true
    connected = uf.connected(5, 0);      // Should be false

    return 0;
}

File: src/1.UnionFind/WeightedQuickUnionUF.py
class WeightedQuickUnionUF:
    def __init__(self, N):
        self.id = list(range(N))
        self.sz = [1] * N

    def _root(self, i):
        while i != self.id[i]:
            # Path compression: make every node in path point to its grandparent
            self.id[i] = self.id[self.id[i]]
            i = self.id[i]
        return i

    def connected(self, p, q):
        return self._root(p) == self._root(q)

    def union(self, p, q):
        i = self._root(p)
        j = self._root(q)
        if i == j:
            return
        # Attach smaller tree to larger tree
        if self.sz[i] < self.sz[j]:
            self.id[i] = j
            self.sz[j] += self.sz[i]
        else:
            self.id[j] = i
            self.sz[i] += self.sz[j]

if __name__ == '__main__':
    uf = WeightedQuickUnionUF(10)
    uf.union(4, 3)
    uf.union(3, 8)
    uf.union(6, 5)
    uf.union(9, 4)
    uf.union(2, 1)
    uf.union(5, 0)
    uf.union(7, 2)
    uf.union(6, 1)
    uf.union(7, 3)
    print(uf.id)  # [6, 2, 6, 4, 6, 6, 6, 2, 4, 4]
    